pkg/k8s/objectcache/objectcache.go

```diff
@@ -149,48 +183,31 @@ func (oc *ObjectCache) Get(ctx context.Context, object util.ObjectRef) (*unstruc
 		ResourceVersion: "0",
 	})
 	if err != nil && !k8serrors.IsNotFound(err) {
-		metric.Error = string(k8serrors.ReasonForError(err))
-		return nil, err
+		return nil, err, string(k8serrors.ReasonForError(err))
 	}
 
 	if err == nil {
-		json, err := raw.MarshalJSON()
-		if err != nil {
-			metric.Error = "Marshal"
-			return nil, fmt.Errorf("server responds with non-marshalable data")
-		}
-
-		err = oc.cache.Set(key, json, int(oc.options.storeTtl.Seconds()))
-		if err != nil {
-			// the object is too large, so just don't cache it
-			metric.Error = "ValueTooLarge"
-		} else {
-			metric.Error = "Penetrated"
-		}
-
-		return raw, nil
+		return raw, nil, "Apiserver"
 	}
-	// else, not found
-
-	metric.Error = "DeletionSnapshot"
 
+	// not found from apiserver, try deletion snapshot instead
 	snapshot, err := oc.DiffCache.FetchSnapshot(ctx, object, diffcache.SnapshotNameDeletion)
 	if err != nil {
-		return nil, metrics.LabelError(fmt.Errorf("cannot fallback to snapshot: %w", err), "SnapshotFetch")
+		return nil, metrics.LabelError(fmt.Errorf("cannot fallback to snapshot: %w", err), "SnapshotFetch"), "SnapshotFetch"
 	}
 
 	if snapshot != nil {
 		uns := &unstructured.Unstructured{}
 		if err := uns.UnmarshalJSON(snapshot.Value); err != nil {
-			return nil, metrics.LabelError(fmt.Errorf("decode snapshot err: %w", err), "SnapshotDecode")
+			return nil, metrics.LabelError(fmt.Errorf("decode snapshot err: %w", err), "SnapshotDecode"), "SnapshotDecode"
 		}
 
-		return uns, nil
+		return uns, nil, "DeletionSnapshot"
 	}
 
 	// all methods failed
 
-	return nil, nil
+	return nil, nil, "NotFoundAnywhere"
```

Comment:
err is nil and value is nil in this case,  value.MarshalJSON will cause nil panic